app/lib/schema_file_loader.rb

```diff
@@ -0,0 +1,109 @@
+
+
+class SchemaFileLoader
+
+  BUCKET = "vita-min-irs-e-file-schema-prod"
+  REGION = "us-east-1"
+  EFILE_SCHEMAS_FILENAMES = [
+    ["efile1040x_2020v5.1.zip", "irs"],
+    ["efile1040x_2021v5.2.zip", "irs"],
+    ["efile1040x_2022v5.3.zip", "irs"],
+    ["efile1040x_2023v5.0.zip", "irs"],
+    ["NYSIndividual2023V4.0.zip", "us_states"],
+    ["AZIndividual2023v1.0.zip", "us_states"],
+  ].freeze
+
+  class << self
+    def load_file(*path)
+      # First we check the vendor directory - if the file is there use it (Non heroku cases)
+      file = File.join(Rails.root, "vendor", *path)
+      return file if File.exist?(file)
+
+      # Next we check the tmp directory in case we already downloaded the file
+      file = File.join(Rails.root, "tmp", *path)
+      download_and_unzip_schemas_from_s3(File.join(Rails.root, "tmp")) unless File.exist?(file)
+
+      file
+    end
+
+    def download_and_unzip_schemas_from_s3(dest_dir)
+      prepare_directories(dest_dir)
+      download_schemas_from_s3(dest_dir)
+      unzip_schemas(dest_dir)
+    end
+
+    def download_schemas_from_s3(dest_dir)
+      s3_client = Aws::S3::Client.new(region: REGION, credentials: s3_credentials)
+      SchemaFileLoader::EFILE_SCHEMAS_FILENAMES.each do |(filename, download_folder)|
+        download_path = File.join(dest_dir, download_folder, filename)
+
+        next if File.exist?(download_path) # If the file already exists, do not re-download.
+
+        s3_client.get_object(
+          response_target: download_path,
+          bucket: BUCKET,
+          key: filename,
+          )
+      end
+    end
+
+    def s3_credentials
+      # On Circle CI, get AWS credentials from environment.
+      # In staging, demo, heroku, and prod environment, get credentials from Rails credentials.
+      #
+      # In development, download the file manually from S3. This allows us to avoid storing any AWS credentials in the development secrets.
+      if ENV["AWS_ACCESS_KEY_ID"].present?
+        return Aws::Credentials.new(ENV["AWS_ACCESS_KEY_ID"], ENV["AWS_SECRET_ACCESS_KEY"])
+      end
+
+      Aws::Credentials.new(
+        Rails.application.credentials.dig(:aws, :access_key_id),
+        Rails.application.credentials.dig(:aws, :secret_access_key),
+        )
+    end
+
+    def prepare_directories(dest_dir)
+      [File.join(dest_dir, 'irs', 'unpacked'), File.join(dest_dir, 'us_states', 'unpacked')].each do |unpack_path|
```

Comment:
i wonder if this could get a flattened list of directory names from `EFILE_SCHEMAS_FILENAMES` so we wouldn't have to update it whenever a new directory was added to the list

Reply:
Done!